File: app/src/main/java/com/privat/contacts/data/cloud/model/SubscriptionNet.java
package com.privat.contacts.data.cloud.model;

import com.google.gson.annotations.SerializedName;

import org.jetbrains.annotations.NotNull;

public class SubscriptionNet {
    @NotNull
    private final String plan;
    @NotNull
    private final String status;
    @SerializedName("payment_method")
    @NotNull
    private final String paymentMethod;
    @NotNull
    private final String term;

    public SubscriptionNet(@NotNull String plan,
                           @NotNull String status,
                           @NotNull String paymentMethod,
                           @NotNull String term) {
        this.plan = plan;
        this.status = status;
        this.paymentMethod = paymentMethod;
        this.term = term;
    }

    public <T> T map(int userId, Mapper<T> mapper) {
        return mapper.map(userId, plan,
                status,
                paymentMethod,
                term);
    }

    public interface Mapper<T> {
        T map(int userId,
              @NotNull String plan,
              @NotNull String status,
              @NotNull String paymentMethod,
              @NotNull String term);
    }
}

File: app/src/main/java/com/privat/contacts/data/cloud/model/SubscriptionNet.kt
package com.privat.contacts.data.cloud.model

import com.google.gson.annotations.SerializedName

class SubscriptionNet(private val plan: String,
                      private val status: String,
                      @field:SerializedName("payment_method") private val paymentMethod: String,
                      private val term: String) {
    fun <T> map(userId: Int, mapper: Mapper<T>): T {
        return mapper.map(userId, plan,
                status,
                paymentMethod,
                term)
    }

    interface Mapper<T> {
        fun map(userId: Int,
                plan: String,
                status: String,
                paymentMethod: String,
                term: String): T
    }
}
